src/share/man/man8/compat_openbsd.8
.\"	$MirOS: src/share/man/man8/compat_openbsd.8,v 1.1.7.1 2005/03/06 16:33:41 tg Exp $
.\"	$OpenBSD: compat_freebsd.8,v 1.17 2003/10/25 07:50:51 jmc Exp $
.\"	$NetBSD: compat_linux.8,v 1.1 1995/03/05 23:30:36 fvdl Exp $
.\"
.\" Copyright (c) 1995 Frank van der Linden
.\" All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. All advertising materials mentioning features or use of this software
.\"    must display the following acknowledgement:
.\"      This product includes software developed for the NetBSD Project
.\"      by Frank van der Linden
.\" 4. The name of the author may not be used to endorse or promote products
.\"    derived from this software without specific prior written permission
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
.\" IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
.\" OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
.\" IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT,
.\" INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
.\" NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
.\" DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
.\" THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
.\" (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
.\" THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
.\"
.Dd October 26, 2004
.Dt COMPAT_OPENBSD 8
.Os
.Sh NAME
.Nm compat_openbsd
.Nd setup for running
.Ox
binaries under emulation
.Sh DESCRIPTION
.Mx
supports running
.Ox
binaries.
Most binaries for
.Ox 3.8
and below should work.
.Pp
The
.Ox
compatibility feature is active for kernels compiled
with the
.Nm COMPAT_OPENBSD
option and the kern.emul.openbsd
.Xr sysctl 8
enabled.
.Pp
A lot of programs are dynamically linked.
This means that the
.Ox
shared libraries that the programs depend on and the runtime linker
are also needed.
Additionally, a "shadow root" directory for
.Ox
binaries on the
.Mx
system will have to be created.
This directory is named
.Pa /emul/openbsd .
Any file operations done by
.Ox
programs run under
.Mx
will look in this directory first.
So, if an
.Ox
program opens, for example,
.Pa /etc/passwd ,
.Mx
will first try to open
.Pa /emul/openbsd/etc/passwd ,
and if that does not exist
open the 'real'
.Pa /etc/passwd file .
It is recommended that
.Ox
packages that include configuration files, etc., be installed under
.Pa /emul/openbsd ,
to avoid naming conflicts with possible
.Mx
counterparts.
Shared libraries should also be installed in the shadow tree.
.Pp
Generally, it will only be necessary to look for the shared libraries that
.Ox
binaries depend on the first few times that
.Ox
programs are installed on the
.Mx
system.
After a while, there will be a sufficient set of
.Ox
shared libraries on the system to be able to run newly imported
.Ox
binaries without any extra work.
.Ss Setting up shared libraries
How to get to know which shared libraries
.Ox
binaries need, and where to get them?
Basically, there are two possibilities.
(When following these instructions, root privileges are required on the
.Ox
system to perform the necessary installation steps).
.Bl -enum
.It
Access to an
.Ox
system:
In this case temporarily install the binary there, see what shared
libraries it needs, and copy them to the
.Mx
system.
Example: ftp the
.Ox
binary of SimCity.
Put it on the
.Ox
system, and check which shared libraries it needs by running
'ldd sim':
.Pp
.Dl me@freebsd% ldd /usr/local/lib/SimCity/res/sim
.Bd -unfilled -offset indent
/usr/local/lib/SimCity/res/sim:
	-lXext.6 => /usr/X11R6/lib/libXext.so.6.0 (0x100c1000)
	-lX11.6 => /usr/X11R6/lib/libX11.so.6.0 (0x100c9000)
	-lc.2 => /usr/lib/libc.so.2.1 (0x10144000)
	-lm.2 => /usr/lib/libm.so.2.0 (0x101a7000)
	-lgcc.261 => /usr/lib/libgcc.so.261.0 (0x101bf000)
.Ed
.Pp
All the files from the last column should be placed under
.Pa /emul/openbsd .
The following files would therefore be required on the
.Mx
system:
.Bd -unfilled -offset indent
/emul/openbsd/usr/X11R6/lib/libXext.so.6.0
/emul/openbsd/usr/X11R6/lib/libX11.so.6.0
/emul/openbsd/usr/lib/libc.so.2.1
/emul/openbsd/usr/lib/libm.so.2.0
/emul/openbsd/usr/lib/libgcc.so.261.0
.Ed
.Pp
Note that if an
.Ox
shared library with a matching major revision number to the first
column of the 'ldd' output is already present, it isn't necessary to copy
the file named in the last column to the
.Mx
system; the one already there should work.
It is advisable to copy the shared library anyway, if it is a newer version.
The old one can be removed.
So, if this library exists on the system:
.Pp
.D1 /emul/openbsd/usr/lib/libc.so.2.0
.Pp
and the ldd output for a new binary is:
.Pp
.D1 -lc.2 => /usr/lib/libc.so.2.1 (0x10144000)
.Pp
it isn't necessary to copy
.Pa /usr/lib/libc.so.2.1
too,
because the program should work fine with the slightly older version.
.Pa libc.so
can be replaced anyway, and that should leave:
.Pp
.D1 /emul/openbsd/usr/lib/libc.so.2.1
.Pp
Finally, the
.Ox
runtime linker and its config files must be present on the system.
These files should be copied from the
.Ox
system to their appropriate place on the
.Mx
system (in the
.Pa /emul/openbsd
tree):
.Bd -unfilled -offset indent
usr/libexec/ld.so
var/run/ld.so.hints
.Ed
.It
No access to an
.Ox
system:
In that case, get the extra files from various ftp sites.
Information on where to look for the various files is appended
below.
.Pp
Retrieve the following files (from _one_ ftp site to avoid
any version mismatches), and install them under
.Pa /emul/openbsd
(i.e.
.Pa foo/bar
is installed as
.Pa /emul/openbsd/foo/bar ) :
.Bd -unfilled -offset indent
sbin/ldconfig
usr/bin/ldd
usr/lib/libc.so.x.y
usr/libexec/ld.so
.Ed
.Pp
ldconfig and ldd don't necessarily need to be under
.Pa /emul/openbsd ,
they can be installed elsewhere in the system too.
Just make sure they don't conflict with their native
.Mx
counterparts.
A good idea would be to install them in
.Pa /usr/local/bin
as ldconfig-openbsd and ldd-openbsd or adjust the environment variable
.Ev PATH
appropriately to include /emul/openbsd/{{,usr.}{,s}bin}.
.Pp
Run the
.Ox
ldconfig program with directory arguments in which the
.Ox
runtime linker should look for shared libs.
.Pa /usr/lib
is standard.
For example:
.Bd -literal -offset indent
me@openbsd% mkdir -p /emul/openbsd/var/run
me@openbsd% touch /emul/freebsd/var/run/ld.so.hints
me@openbsd% ldconfig-openbsd /usr/X11R6/lib /usr/local/lib
.Ed
.Pp
Note that argument directories of ldconfig are
mapped to
.Pa /emul/openbsd/XXXX
by MirOS BSD's compat code, and should exist as such on the
.Mx
system.
Make sure
.Pa /emul/openbsd/var/run/ld.so.hints
exists when running FreeBSD's ldconfig; if not, MirOS BSD's
.Pa /var/run/ld.so.hints
could be lost.
.Pp
.Ox
ldconfig should be statically
linked, so it doesn't need any shared libraries by itself.
It will create the file
.Pa /emul/openbsd/var/run/ld.so.hints .
The
.Ox
version of the ldconfig program should be rerun each time a new shared
library is added.
.Pp
The
.Mx
system should now be set up for
.Ox
binaries which only need a shared libc.
Test this by running the
.Ox
ldd on itself.
Suppose that it is installed as ldd-openbsd, it should produce
something like:
.Pp
.Dl me@mirbsd% ldd-openbsd $(which ldd-openbsd)
.Bd -unfilled -offset indent
/usr/local/bin/ldd-freebsd:
	-lc.2 => /usr/lib/libc.so.2.1 (0x1001a000)
.Ed
.Pp
This being done, new
.Ox
binaries can now be installed.
Whenever a new
.Ox
program is installed, it should be determined if it needs shared libraries,
and if so, whether they are installed in the
.Pa /emul/openbsd
tree.
To do this, run the
.Ox
version ldd on the new program, and watch its output.
ldd (see also the manual page for
.Xr ldd 1 )
will print a list of shared libraries that the program depends on, in the
form -l<majorname> => <fullname>.
.Pp
If it prints "not found" instead of <fullname> it means that
an extra library is needed.
Which library this is, is shown
in <majorname>, which will be of the form XXXX.<N>.
Find a libXXXX.so.<N>.<mm> on an
.Ox
ftp site, and install it on the
.Mx
system.
The XXXX (name) and <N> (major revision number) should match;
the minor number(s) <mm> are less important,
though it is advised to take the most recent version.
.El
.Ss Finding the necessary files
.Sy Note :
the information below is valid as of Oct 2004, but certain details
such as names of ftp sites, directories and distribution names
may have changed since then.
.Pp
.Ox
distribution is available on a lot of ftp sites.
Sometimes the files are unpacked,
and individual files can be retrieved, but mostly they
are stored in distribution sets, usually consisting of subdirectories
with gzipped tar files in them.
The primary ftp site for the distribution is:
.Pp
.D1 Pa ftp.openbsd.org:/pub/OpenBSD
.Pp
Mirror sites are described at:
.Pp
.D1 Pa www.openbsd.org/ftp.html
.Pp
This distribution consists of a number of tar-ed and gzipped files.
Normally, they're controlled by an install program, but the files can
be retrieved "by hand" too.
The way to look something up is to retrieve all
the files in the distribution, and "tar ztvf" through them for the files
needed.
Here is an example of a list of files that might be needed:
.Bd -unfilled -offset indent
Needed          Files

ld.so           5.0-RELEASE/base/base.??
ldconfig        5.0-RELEASE/base/base.??
ldd             5.0-RELEASE/base/base.??
libc.so.6       5.0-RELEASE/base/base.??
libX11.so.6     5.0-RELEASE/packages/x11/XFree86-libraries-??
libXt.so.6      5.0-RELEASE/packages/x11/XFree86-libraries-??
.Ed
.Pp
The files called "base.??" are tar-ed, gzipped and split,
so they can be extracted with "cat base.?? | tar zpxf -".
.Pp
The XFree86 libraries are compressed with bzip2 and can be
extracted with "bzcat <file> | tar pxf -".
Note: The bzip2 utilities are not part of a base
.Mx
system.
bzip2 can be added via
.Xr packages 7
or
.Xr ports 7 .
.Pp
Simply extract the files from these compressed tarfiles in the
.Pa /emul/openbsd
directory (possibly omitting or afterwards removing unnecessary files).
.Sh CAVEATS
.Nm compat_openbsd
only works on the i386 architecture.
.Sh BUGS
The information about
.Ox
releases may become outdated.
This manual page is too visibly derived from the freebsd one.


src/sys/compat/openbsd/compat_openbsd.c
/**	$MirOS: src/sys/compat/openbsd/compat_openbsd.c,v 1.1.7.1 2005/03/06 16:33:43 tg Exp $ */
/* 	$OpenBSD: compat_aout.c,v 1.1 2003/06/24 22:45:33 espie Exp $ */

/*
 * Copyright (c) 2004
 *	Thorsten "mirabile" Glaser <tg@66h.42h.de>
 * Copyright (c) 2003 Marc Espie
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. The name of the author may not be used to endorse or promote products
 *    derived from this software without specific prior written permission
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
 * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
 * OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
 * IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT,
 * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
 * NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
 * THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

#include <sys/param.h>
#include <sys/syscall.h>
#include <sys/signalvar.h>
#include <sys/mount.h>
#include <sys/fcntl.h>
#include <sys/exec.h>
#include <sys/exec_olf.h>
#include <sys/malloc.h>
#include <compat/common/compat_util.h>
#include <compat/openbsd/compat_openbsd.h>
#include <compat/openbsd/openbsd_syscall.h>
#include <compat/openbsd/openbsd_syscallargs.h>

void openbsd_compat_setup(struct exec_package *epp);

extern char sigcode[], esigcode[];
extern struct sysent openbsd_sysent[];
#ifdef SYSCALL_DEBUG
extern char *openbsd_syscallnames[];
#endif

#ifdef __sparc__
#define DEFAULT_EMULFLAGS	EMUL_ENABLED
#else
#define DEFAULT_EMULFLAGS	0
#endif

struct emul emul_openbsd_aout = {
	"openbsd",
	NULL,
	sendsig,
	OPENBSD_SYS_syscall,
	OPENBSD_SYS_MAXSYSCALL,
	openbsd_sysent,
#ifdef SYSCALL_DEBUG
	openbsd_syscallnames,
#else
	NULL,
#endif
	0,
	copyargs,
	setregs,
	NULL,
	sigcode,
	esigcode,
	DEFAULT_EMULFLAGS
};

struct emul emul_openbsd = {
	"openbsd",
	NULL,
	sendsig,
	OPENBSD_SYS_syscall,
	OPENBSD_SYS_MAXSYSCALL,
	openbsd_sysent,
#ifdef SYSCALL_DEBUG
	openbsd_syscallnames,
#else
	NULL,
#endif
	sizeof(AuxInfo) * ELF_AUX_ENTRIES,
	elf32_copyargs,
	setregs,
	exec_elf32_fixup,
	sigcode,
	esigcode,
	DEFAULT_EMULFLAGS
};

const char openbsd_emul_path[] = "/emul/openbsd";

void
openbsd_compat_setup(struct exec_package *epp)
{
	epp->ep_emul = &emul_openbsd;
}

int
exec_openbsd_elf32_makecmds(struct proc *p, struct exec_package *epp)
{
	if (!(emul_openbsd.e_flags & EMUL_ENABLED))
		return (ENOEXEC);
	return exec_elf32_makecmds(p, epp);
}

int
openbsd_elf_probe(struct proc *p, struct exec_package *epp,
    char *itp, u_long *pos, u_int8_t *os)
{
	char *bp;
	int error;
	size_t len;

	if (!elf32_os_pt_note(p, epp, epp->ep_hdr, "MirOS OpenBSD", 14, 4))
		goto emulated;

	if (!elf32_os_pt_note(p, epp, epp->ep_hdr, "OpenBSD", 8, 4))
		goto emulated;

	return ENOEXEC;

emulated:
	if (itp[0]) {
		if ((error = emul_find(p, NULL, openbsd_emul_path, itp, &bp, 0)))
			return (error);
		if ((error = copystr(bp, itp, MAXPATHLEN, &len)))
			return (error);
		free(bp, M_TEMP);
	}
	epp->ep_emul = &emul_openbsd;
	*pos = ELF32_NO_ADDR;
	if (*os == OOS_NULL)
		*os = OOS_OPENBSD;
	return (0);
}

#include <sys/sysctl.h>

int
openbsd_sys___sysctl(struct proc *p, void *v, register_t *retval)
{
	extern int sys___sysctl(struct proc *, void *, register_t *);
	struct openbsd_sys___sysctl_args /* {
		syscallarg(int *) name;
		syscallarg(u_int) namelen;
		syscallarg(void *) old;
		syscallarg(size_t *) oldlenp;
		syscallarg(void *) new;
		syscallarg(size_t) newlen;
	} */ *uap = v;
	struct sys___sysctl_args {
		syscallarg(int *) name;
		syscallarg(u_int) namelen;
		syscallarg(void *) old;
		syscallarg(size_t *) oldlenp;
		syscallarg(void *) new;
		syscallarg(size_t) newlen;
	} bap;
	int error;
	int name[CTL_MAXNAME], *namep;
	caddr_t sg;

	if (SCARG(uap, new) != NULL &&
	    (error = suser(p, 0)))
		return (error);
	/*
	 * all top-level sysctl names are non-terminal
	 */
	if (SCARG(uap, namelen) > CTL_MAXNAME || SCARG(uap, namelen) < 2)
		return EINVAL;
	if ((error = copyin(SCARG(uap, name), name,
	    SCARG(uap, namelen) * sizeof(int))))
		return error;

	sg = stackgap_init(p->p_emul);
	namep = stackgap_alloc(&sg, sizeof(name));

	/*
	 * XXX Translation OpenBSD -> MirOS sysctl numbers
	 */
	if ((name[0] == CTL_KERN) && (name[1] == KERN_OSTYPE)
	    && (SCARG(uap, new) == NULL)) {
		name[1] = KERN_EMULUNAME;
	}

	if ((error = copyout(name, namep, sizeof(name))))
		return error;

	SCARG(&bap, name) = namep;
	SCARG(&bap, namelen) = SCARG(uap, namelen);
	SCARG(&bap, old) = SCARG(uap, old);
	SCARG(&bap, oldlenp) = SCARG(uap, oldlenp);
	SCARG(&bap, new) = SCARG(uap, new);
	SCARG(&bap, newlen) = SCARG(uap, newlen);

	return sys___sysctl(p, &bap, retval);
}
